File: vault/vault.go
package vault

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"github.com/wuedmo/lazypassword/crypto"
	"github.com/wuedmo/lazypassword/entry"
)

// VaultData is the structure stored in the encrypted vault file
type VaultData struct {
	Entries []entry.Entry `json:"entries"`
}

// Vault manages the encrypted password vault
type Vault struct {
	path      string
	data      VaultData
	salt      []byte
	key       []byte
	locked    bool
	modified  bool
}

// NewVault creates a new vault instance (does not create file)
func NewVault(path string) *Vault {
	return &Vault{
		path:     path,
		data:     VaultData{Entries: []entry.Entry{}},
		salt:     nil,
		key:      nil,
		locked:   true,
		modified: false,
	}
}

// Create initializes a new vault file with the given password
// Creates: [salt:32][encrypted_vault_data]
func (v *Vault) Create(password string) error {
	if !v.locked {
		return fmt.Errorf("vault is already unlocked")
	}

	// Generate salt and derive key
	v.salt = crypto.GenerateSalt()
	v.key = crypto.DeriveKey(password, v.salt)

	// Initialize empty vault data
	v.data = VaultData{Entries: []entry.Entry{}}
	v.locked = false
	v.modified = true

	// Save to disk
	return v.Save()
}

// Unlock decrypts and loads the vault with the given password
func (v *Vault) Unlock(password string) error {
	if !v.locked {
		return fmt.Errorf("vault is already unlocked")
	}

	// Read vault file
	data, err := os.ReadFile(v.path)
	if err != nil {
		return fmt.Errorf("failed to read vault file: %w", err)
	}

	if len(data) < 32 {
		return fmt.Errorf("vault file too short")
	}

	// Extract salt and encrypted data
	v.salt = make([]byte, 32)
	copy(v.salt, data[:32])
	encryptedData := data[32:]

	// Derive key
	v.key = crypto.DeriveKey(password, v.salt)

	// Decrypt
	plaintext, err := crypto.Decrypt(encryptedData, v.key)
	if err != nil {
		// Wipe key on decryption failure
		crypto.SecureWipe(v.key)
		v.key = nil
		v.salt = nil
		return fmt.Errorf("failed to decrypt vault: %w", err)
	}

	// Parse JSON
	var vaultData VaultData
	if err := json.Unmarshal(plaintext, &vaultData); err != nil {
		crypto.SecureWipe(v.key)
		v.key = nil
		v.salt = nil
		return fmt.Errorf("failed to parse vault data: %w", err)
	}

	v.data = vaultData
	v.locked = false
	v.modified = false

	// Wipe plaintext from memory
	crypto.SecureWipe(plaintext)

	return nil
}

// Lock wipes decrypted data from memory
func (v *Vault) Lock() {
	if v.key != nil {
		crypto.SecureWipe(v.key)
		v.key = nil
	}
	if v.salt != nil {
		// Salt is not secret, but clear for hygiene
		v.salt = nil
	}
	v.data = VaultData{Entries: []entry.Entry{}}
	v.locked = true
	v.modified = false
}

// IsLocked returns true if the vault is currently locked
func (v *Vault) IsLocked() bool {
	return v.locked
}

// AddEntry adds a new entry to the vault
func (v *Vault) AddEntry(e entry.Entry) error {
	if v.locked {
		return fmt.Errorf("vault is locked")
	}

	// Check for duplicate ID
	for _, existing := range v.data.Entries {
		if existing.ID == e.ID {
			return fmt.Errorf("entry with ID %s already exists", e.ID)
		}
	}

	v.data.Entries = append(v.data.Entries, e)
	v.modified = true
	return nil
}

// UpdateEntry updates an existing entry by ID
func (v *Vault) UpdateEntry(id string, updated entry.Entry) error {
	if v.locked {
		return fmt.Errorf("vault is locked")
	}

	for i, e := range v.data.Entries {
		if e.ID == id {
			// Preserve creation timestamp
			updated.ID = id
			updated.CreatedAt = e.CreatedAt
			v.data.Entries[i] = updated
			v.modified = true
			return nil
		}
	}

	return fmt.Errorf("entry with ID %s not found", id)
}

// DeleteEntry removes an entry by ID
func (v *Vault) DeleteEntry(id string) error {
	if v.locked {
		return fmt.Errorf("vault is locked")
	}

	for i, e := range v.data.Entries {
		if e.ID == id {
			// Remove entry
			v.data.Entries = append(v.data.Entries[:i], v.data.Entries[i+1:]...)
			v.modified = true
			return nil
		}
	}

	return fmt.Errorf("entry with ID %s not found", id)
}

// GetEntries returns all entries (slice copy)
func (v *Vault) GetEntries() ([]entry.Entry, error) {
	if v.locked {
		return nil, fmt.Errorf("vault is locked")
	}

	// Return a copy to prevent external modification
	entries := make([]entry.Entry, len(v.data.Entries))
	copy(entries, v.data.Entries)
	return entries, nil
}

// Save encrypts and writes the vault atomically
func (v *Vault) Save() error {
	if v.locked {
		return fmt.Errorf("vault is locked")
	}

	if v.key == nil {
		return fmt.Errorf("no key available")
	}

	// Marshal vault data
	plaintext, err := json.Marshal(v.data)
	if err != nil {
		return fmt.Errorf("failed to marshal vault data: %w", err)
	}

	// Encrypt
	ciphertext, err := crypto.Encrypt(plaintext, v.key)
	if err != nil {
		return fmt.Errorf("failed to encrypt vault: %w", err)
	}

	// Build final data: [salt][ciphertext]
	data := make([]byte, 32+len(ciphertext))
	copy(data, v.salt)
	copy(data[32:], ciphertext)

	// Atomic write using temp file + rename
	dir := filepath.Dir(v.path)
	if err := os.MkdirAll(dir, 0700); err != nil {
		return fmt.Errorf("failed to create directory: %w", err)
	}

	tmpFile, err := os.CreateTemp(dir, ".lazypassword-*.tmp")
	if err != nil {
		return fmt.Errorf("failed to create temp file: %w", err)
	}

	tmpPath := tmpFile.Name()
	defer os.Remove(tmpPath)

	if _, err := tmpFile.Write(data); err != nil {
		tmpFile.Close()
		return fmt.Errorf("failed to write temp file: %w", err)
	}

	if err := tmpFile.Chmod(0600); err != nil {
		tmpFile.Close()
		return fmt.Errorf("failed to set permissions: %w", err)
	}

	if err := tmpFile.Close(); err != nil {
		return fmt.Errorf("failed to close temp file: %w", err)
	}

	// Atomic rename
	if err := os.Rename(tmpPath, v.path); err != nil {
		return fmt.Errorf("failed to rename temp file: %w", err)
	}

	v.modified = false

	// Wipe plaintext from memory
	crypto.SecureWipe(plaintext)

	return nil
}

// IsModified returns true if the vault has unsaved changes
func (v *Vault) IsModified() bool {
	return v.modified
}

// GetEntryByID returns a single entry by ID
func (v *Vault) GetEntryByID(id string) (*entry.Entry, error) {
	if v.locked {
		return nil, fmt.Errorf("vault is locked")
	}

	for _, e := range v.data.Entries {
		if e.ID == id {
			copy := e
			return &copy, nil
		}
	}

	return nil, fmt.Errorf("entry with ID %s not found", id)
}

// SearchEntries returns entries matching a query (case-insensitive)
func (v *Vault) SearchEntries(query string) ([]entry.Entry, error) {
	if v.locked {
		return nil, fmt.Errorf("vault is locked")
	}

	// Simple search - can be expanded
	queryLower := []byte(query)
	for i := range queryLower {
		if queryLower[i] >= 'A' && queryLower[i] <= 'Z' {
			queryLower[i] += 32
		}
	}

	var results []entry.Entry
	for _, e := range v.data.Entries {
		if containsIgnoreCase(e.Title, query) ||
			containsIgnoreCase(e.Username, query) ||
			containsIgnoreCase(e.URL, query) ||
			containsIgnoreCase(e.Notes, query) {
			results = append(results, e)
		}
	}

	return results, nil
}

func containsIgnoreCase(s, substr string) bool {
	if len(substr) == 0 {
		return true
	}
	if len(s) < len(substr) {
		return false
	}

	// Simple case-insensitive check
	lowerS := []byte(s)
	for i := range lowerS {
		if lowerS[i] >= 'A' && lowerS[i] <= 'Z' {
			lowerS[i] += 32
		}
	}

	lowerSubstr := []byte(substr)
	for i := range lowerSubstr {
		if lowerSubstr[i] >= 'A' && lowerSubstr[i] <= 'Z' {
			lowerSubstr[i] += 32
		}
	}

	// Simple contains
	sl, sub := string(lowerS), string(lowerSubstr)
	for i := 0; i <= len(sl)-len(sub); i++ {
		if sl[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}

File: lazypassword/vault.py
"""
Vault management for lazypassword.
Handles encrypted storage of password entries with atomic writes.
"""

import json
import os
import shutil
import tempfile
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

from cryptography.exceptions import InvalidTag

from lazypassword.crypto import (
    derive_key,
    encrypt,
    decrypt,
    generate_salt,
    secure_wipe,
)


VAULT_VERSION = 1
NONCE_SALT_SIZE = 32  # Combined salt + encrypted data


class VaultError(Exception):
    """Base exception for vault operations."""
    pass


class VaultCorruptedError(VaultError):
    """Raised when vault file is corrupted or invalid."""
    pass


class VaultAuthError(VaultError):
    """Raised when password is incorrect."""
    pass


class Vault:
    """
    Encrypted vault for storing password entries.
    
    The vault file format:
    - First 32 bytes: salt for key derivation
    - Remaining bytes: encrypted JSON data (nonce + ciphertext + tag)
    """
    
    def __init__(self, vault_path: str):
        """
        Initialize vault instance.
        
        Args:
            vault_path: Path to the vault file
        """
        self.vault_path = Path(vault_path)
        self._key: Optional[bytes] = None
        self._salt: Optional[bytes] = None
        self._data: dict = {}
        self._locked = True
    
    def create(self, password: str) -> None:
        """
        Create a new vault with encrypted empty structure.
        
        Args:
            password: Master password for the vault
        """
        self._salt = generate_salt()
        self._key = derive_key(password, self._salt)
        
        # Initialize empty vault structure
        self._data = {
            "version": VAULT_VERSION,
            "entries": [],
            "settings": {
                "auto_lock_timeout": 300,  # 5 minutes
                "clipboard_clear_delay": 30,  # 30 seconds
                "theme": "dark",  # default theme
            }
        }
        
        self._locked = False
        self.save()
    
    def unlock(self, password: str) -> bool:
        """
        Unlock the vault with the provided password.
        
        Args:
            password: Master password to attempt
            
        Returns:
            bool: True if unlock succeeded, False otherwise
        """
        if not self.vault_path.exists():
            return False
        
        try:
            with open(self.vault_path, 'rb') as f:
                file_content = f.read()
            
            if len(file_content) < NONCE_SALT_SIZE + 28:
                raise VaultCorruptedError("Vault file is too small or corrupted")
            
            # Extract salt and encrypted data
            salt = file_content[:NONCE_SALT_SIZE]
            encrypted_data = file_content[NONCE_SALT_SIZE:]
            
            # Derive key and attempt decryption
            key = derive_key(password, salt)
            
            try:
                decrypted = decrypt(encrypted_data, key)
                vault_data = json.loads(decrypted.decode('utf-8'))
                
                # Validate vault structure
                if "version" not in vault_data or "entries" not in vault_data:
                    secure_wipe(bytearray(key))
                    raise VaultCorruptedError("Invalid vault structure")
                
                # Success - store unlocked state
                self._salt = salt
                self._key = key
                self._data = vault_data
                self._locked = False
                return True
                
            except (ValueError, InvalidTag) as e:
                # Decryption failed - wrong password or corrupted
                secure_wipe(bytearray(key))
                return False
                
        except (IOError, OSError) as e:
            raise VaultError(f"Failed to read vault file: {e}")
        except json.JSONDecodeError as e:
            raise VaultCorruptedError(f"Invalid vault data: {e}")
    
    def lock(self) -> None:
        """Lock the vault and securely wipe decrypted data from memory."""
        if self._key:
            secure_wipe(bytearray(self._key))
            self._key = None
        
        self._data = {}
        self._locked = True
    
    def is_locked(self) -> bool:
        """Check if the vault is currently locked."""
        return self._locked
    
    def _ensure_unlocked(self) -> None:
        """Raise error if vault is locked."""
        if self._locked:
            raise VaultError("Vault is locked. Unlock first.")
    
    def _generate_entry_id(self) -> str:
        """Generate a unique entry ID."""
        return str(uuid.uuid4())
    
    def _get_timestamp(self) -> str:
        """Get current UTC timestamp in ISO format."""
        return datetime.now(timezone.utc).isoformat()
    
    def add_entry(self, entry: dict) -> str:
        """
        Add a new entry to the vault.
        
        Args:
            entry: Dictionary with entry data (title, username, password, etc.)
            
        Returns:
            str: The generated entry ID
        """
        self._ensure_unlocked()
        
        entry_id = self._generate_entry_id()
        timestamp = self._get_timestamp()
        
        new_entry = {
            "id": entry_id,
            "created_at": timestamp,
            "updated_at": timestamp,
            **entry
        }
        
        self._data["entries"].append(new_entry)
        return entry_id
    
    def update_entry(self, entry_id: str, entry: dict) -> None:
        """
        Update an existing entry.
        
        Args:
            entry_id: ID of the entry to update
            entry: Dictionary with updated fields
            
        Raises:
            VaultError: If entry not found
        """
        self._ensure_unlocked()
        
        for i, existing in enumerate(self._data["entries"]):
            if existing.get("id") == entry_id:
                # Merge updates, preserve id and created_at
                updated = {
                    **existing,
                    **entry,
                    "id": entry_id,  # Prevent ID change
                    "created_at": existing["created_at"],
                    "updated_at": self._get_timestamp(),
                }
                self._data["entries"][i] = updated
                return
        
        raise VaultError(f"Entry with ID '{entry_id}' not found")
    
    def delete_entry(self, entry_id: str) -> None:
        """
        Delete an entry from the vault.
        
        Args:
            entry_id: ID of the entry to delete
            
        Raises:
            VaultError: If entry not found
        """
        self._ensure_unlocked()
        
        for i, entry in enumerate(self._data["entries"]):
            if entry.get("id") == entry_id:
                del self._data["entries"][i]
                return
        
        raise VaultError(f"Entry with ID '{entry_id}' not found")
    
    def get_entries(self) -> list:
        """
        Get all entries from the vault.
        
        Returns:
            list: List of entry dictionaries
        """
        self._ensure_unlocked()
        return list(self._data.get("entries", []))
    
    def search(self, query: str) -> list:
        """
        Search entries by query string.
        
        Args:
            query: Search string (case-insensitive)
            
        Returns:
            list: Matching entries
        """
        self._ensure_unlocked()
        
        if not query:
            return self.get_entries()
        
        query_lower = query.lower()
        results = []
        
        for entry in self._data.get("entries", []):
            # Search in title, username, url, notes
            searchable_fields = [
                entry.get("title", ""),
                entry.get("username", ""),
                entry.get("url", ""),
                entry.get("notes", ""),
            ]
            
            if any(query_lower in field.lower() for field in searchable_fields):
                results.append(entry)
        
        return results
    
    def get_setting(self, key: str, default: Any = None) -> Any:
        """
        Get a setting value from the vault.
        
        Args:
            key: Setting key
            default: Default value if key not found
            
        Returns:
            Setting value or default
        """
        self._ensure_unlocked()
        return self._data.get("settings", {}).get(key, default)
    
    def set_setting(self, key: str, value: Any) -> None:
        """
        Set a setting value in the vault.
        
        Args:
            key: Setting key
            value: Setting value
        """
        self._ensure_unlocked()
        if "settings" not in self._data:
            self._data["settings"] = {}
        self._data["settings"][key] = value
    
    def get_theme(self) -> str:
        """Get the current theme name."""
        return self.get_setting("theme", "dark")
    
    def set_theme(self, theme: str) -> None:
        """
        Set the theme name.
        
        Args:
            theme: Theme name (e.g., 'dark', 'light', 'nord', etc.)
        """
        self.set_setting("theme", theme)
    
    def save(self) -> None:
        """
        Save the vault to disk with atomic write and backup.
        """
        self._ensure_unlocked()
        
        if not self._key or not self._salt:
            raise VaultError("Cannot save: vault not properly initialized")
        
        # Serialize and encrypt
        json_data = json.dumps(self._data, separators=(',', ':')).encode('utf-8')
        encrypted = encrypt(json_data, self._key)
        
        # Combine salt + encrypted data
        file_content = self._salt + encrypted
        
        # Ensure parent directory exists
        self.vault_path.parent.mkdir(parents=True, exist_ok=True)
        
        # Create backup if file exists
        if self.vault_path.exists():
            backup_path = self.vault_path.with_suffix('.vault.bak')
            try:
                shutil.copy2(self.vault_path, backup_path)
            except (IOError, OSError):
                pass  # Best effort backup
        
        # Atomic write: write to temp file, then rename
        try:
            fd, temp_path = tempfile.mkstemp(
                dir=self.vault_path.parent,
                prefix='.vault_tmp_'
            )
            try:
                os.write(fd, file_content)
                os.fsync(fd)  # Ensure data is written to disk
            finally:
                os.close(fd)
            
            # Atomic rename
            os.replace(temp_path, self.vault_path)
            
        except (IOError, OSError) as e:
            # Clean up temp file if it exists
            if 'temp_path' in locals() and os.path.exists(temp_path):
                os.unlink(temp_path)
            raise VaultError(f"Failed to save vault: {e}")
